lenskit-core/src/main/java/org/grouplens/lenskit/data/event/Like.java

```diff
@@ -20,14 +20,120 @@
  */
 package org.grouplens.lenskit.data.event;
 
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+
+import java.io.Serializable;
+
 /**
  * Generic interface representing a unary interaction.  This provides a basic, generic representation
  * of likes, clicks, purchases, and other unary interactions.  Applications may want to implement
  * their own events to represent such interactions.
+ * <p>
+ * This class can be created with {@link LikeBuilder} or the {@link #create(long, long, long)} method.
+ * </p>
  *
  * @since 2.2
  * @see Events#like(long, long)
  * @see Events#like(long, long, long)
  */
-public interface Like extends Event {
+public final class Like implements Event, Serializable {
+    private static final long serialVersionUID = 1L;
+
+    private final long userId;
+    private final long itemId;
+    private final long timestamp;
+
+    Like(long uid, long iid, long ts) {
+        userId = uid;
+        itemId = iid;
+        timestamp = ts;
+    }
+
+    /**
+     * Create a new like builder.
+     * @return A new builder for Like events.
+     */
+    public static LikeBuilder newBuilder() {
+        return new LikeBuilder();
+    }
+
+    /**
+     * Create a new {@link Like} event with no timestamp.
+     *
+     * @param user The user ID.
+     * @param item The item ID.
+     * @return A {@link Like}.
+     */
+    public static Like create(long user, long item) {
+        return create(user, item, -1);
+    }
+
+    /**
+     * Create a new {@link Like} event.
+     *
+     * @param user The user ID.
+     * @param item The item ID.
+     * @param ts The timestamp.
+     * @return A {@link Like}.
+     */
+    public static Like create(long user, long item, long ts) {
+        return new Like(user, item, ts);
+    }
+
+    @Override
+    public long getUserId() {
+        return userId;
+    }
+
+    @Override
+    public long getItemId() {
+        return itemId;
+    }
+
+    @Override
+    public long getTimestamp() {
+        return timestamp;
+    }
+
+    /**
+     * Create a new builder initialized with a copy of this event's contents.
+     * @return A new builder initialized with this event's data.
+     */
+    public LikeBuilder copyBuilder() {
```

Comment:
the CloneBuilder as a way to "modify" an immutable object argument is compelling, but not intuitive to me. I definitely say we keep the copy builder then. Is there any way to work this idiom into some sort of example project?

don't implement cloneable. Who would want to clone an imitable object with a well defined equality semantics?